// @jlguenego/asn.1/build/src/actions/ActionType.js
"use strict";
/*
BuiltinType ::=
BitStringType
| BooleanType
| CharacterStringType
| ChoiceType
| DateType
| DateTimeType
| DurationType
| EmbeddedPDVType
| EnumeratedType
| ExternalType
| InstanceOfType
| IntegerType
| IRIType
| NullType
| ObjectClassFieldType
| ObjectIdentifierType
| OctetStringType
| RealType
| RelativeIRIType
| RelativeOIDType
| SequenceType
| SequenceOfType
| SetType
| SetOfType
| PrefixedType
| TimeType
| TimeOfDayType
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionType = void 0;
var ActionType;
(function (ActionType) {
    ActionType["NONE"] = "none";
    ActionType["INIT"] = "init";
    ActionType["OBJECT"] = "object";
    ActionType["ITEM"] = "item";
    ActionType["TYPE_BOOLEAN"] = "type boolean";
    ActionType["TYPE_INTEGER"] = "type integer";
    ActionType["TYPE_BIT_STRING"] = "type bit string";
    ActionType["TYPE_OCTET_STRING"] = "type octet string";
    ActionType["TYPE_NULL"] = "type null";
    ActionType["TYPE_OBJECT_IDENTIFER"] = "type Object Identifier";
    ActionType["TYPE_CHARACTERSTRING"] = "type characterstring";
    ActionType["TYPE_CHOICE"] = "type choice";
    ActionType["TYPE_DATE"] = "type date";
    ActionType["TYPE_DATETIME"] = "type datetime";
    ActionType["TYPE_DURATION"] = "type duration";
    ActionType["TYPE_IA5STRING"] = "type IA5string";
    ActionType["TYPE_T61STRING"] = "type T61string";
    ActionType["TYPE_PRINTABLE_STRING"] = "type printable string";
    ActionType["TYPE_UTCTIME"] = "type utc time";
})(ActionType = exports.ActionType || (exports.ActionType = {}));
