// @jlguenego/asn.1/build/src/actions/InitAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitAction = void 0;
const Action_1 = require("../Action");
const ActionType_1 = require("./ActionType");
const SEQUENCE = 0x30;
class InitAction extends Action_1.Action {
    constructor() {
        super(...arguments);
        this.type = ActionType_1.ActionType.INIT;
    }
    transform(state) {
        if (state.index === state.size) {
            state.nextAction = ActionType_1.ActionType.NONE;
            return;
        }
        const type = state.dataview.getUint8(state.index);
        if (type === SEQUENCE) {
            const sequence = {};
            state.context = {
                sequence: sequence,
                length: -1,
                index: 0,
            };
            state.trees.push(sequence);
            state.nextAction = ActionType_1.ActionType.SEQUENCE;
            state.index += 1;
            return;
        }
        throw new Error('Unexpected Type');
    }
}
exports.InitAction = InitAction;
