// @jlguenego/asn.1/build/src/codec/der/decoder/actions/ObjectAction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectAction = void 0;
const debug_1 = __importDefault(require("debug"));
const Action_1 = require("../../../../Action");
const ActionType_1 = require("../../../../actions/ActionType");
const misc_1 = require("../../../ber/decoder/misc");
const debug = debug_1.default('asn.1:sequence');
class ObjectAction extends Action_1.Action {
    constructor() {
        super(...arguments);
        this.type = ActionType_1.ActionType.OBJECT;
    }
    transform(state) {
        if (!state.context) {
            state.context = {
                current: {},
                length: -1,
                index: 0,
            };
            state.root = state.context.current;
        }
        else {
            // sequence inside sequence
            const newSequence = {};
            const key = Object.keys(state.context.current).length;
            state.context.current[key] = newSequence;
            state.context = {
                current: newSequence,
                length: -1,
                index: 0,
                parent: state.context,
            };
        }
        const length = misc_1.readLengthOctets(state);
        debug('length: ', length);
        const context = state.context;
        context.length = length;
        if (length === 0) {
            state.nextAction = ActionType_1.ActionType.INIT;
            return;
        }
        state.nextAction = ActionType_1.ActionType.ITEM;
    }
}
exports.ObjectAction = ObjectAction;
