// @jlguenego/asn.1/build/test/jlg-protocol.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const fs_1 = require("fs");
const path_1 = require("path");
const src_1 = require("../src");
const EncodingRule_1 = require("../src/EncodingRule");
describe('JLG Protocol', () => {
    it('test JLGDER message', () => {
        const inputMsg = fs_1.readFileSync(path_1.resolve(__dirname, 'data/jlg-test.der'), {
            encoding: 'utf8',
        });
        const output = src_1.ASN1.parseMsg(inputMsg, {
            encodingRule: EncodingRule_1.EncodingRule.DER,
            format: 'hex',
        });
        assert_1.default.deepStrictEqual(output, {
            tagClass: 'UNIVERSAL',
            isConstructed: true,
            tagCode: 16,
            tagLabel: 'SEQUENCE',
            length: 14,
            lengthType: 'DEFINITE',
            value: [
                {
                    tagClass: 'UNIVERSAL',
                    isConstructed: false,
                    tagCode: 1,
                    tagLabel: 'BOOLEAN',
                    length: 1,
                    lengthType: 'DEFINITE',
                    value: false,
                },
                {
                    tagClass: 'UNIVERSAL',
                    isConstructed: false,
                    tagCode: 1,
                    tagLabel: 'BOOLEAN',
                    length: 1,
                    lengthType: 'DEFINITE',
                    value: true,
                },
                {
                    tagClass: 'UNIVERSAL',
                    isConstructed: true,
                    tagCode: 16,
                    tagLabel: 'SEQUENCE',
                    length: 6,
                    lengthType: 'DEFINITE',
                    value: [
                        {
                            tagClass: 'UNIVERSAL',
                            isConstructed: false,
                            tagCode: 1,
                            tagLabel: 'BOOLEAN',
                            length: 1,
                            lengthType: 'DEFINITE',
                            value: true,
                        },
                        {
                            tagClass: 'UNIVERSAL',
                            isConstructed: false,
                            tagCode: 1,
                            tagLabel: 'BOOLEAN',
                            length: 1,
                            lengthType: 'DEFINITE',
                            value: true,
                        },
                    ],
                },
            ],
        });
    });
});
